### ISSUE TITLE
Shell module cannot cope with spaces in filenames

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7292]
Migrated from [rt.perl.org#29585](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=29585) (status was 'resolved')

Searchable as RT29585$

**Comment by p5pRT** [comment]
### From swhite@decisionsoft.com

#### Created by swhite@decisionsoft.com



The Shell module can't cope with spaces in filenames\.  Take the following code

use Shell qw\(rm\);

rm\("my file"\);


It doesn't work\.  In a fit of cunningness you try&#8203;:

rm\("'my file'"\);

But the Shell module nicely escapes the single quotes for you so that doesn't work either\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.8.0:

Configured by bhcompile at Sun Sep  1 23:55:07 EDT 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.18-11smp, archname=i386-linux-thread-multi
    uname='linux daffy.perf.redhat.com 2.4.18-11smp #1 smp thu aug 15 06:41:59 edt 2002 i686 i686 i386 gnulinux '
    config_args='-des -Doptimize=-O2 -march=i386 -mcpu=i686 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Darchname=i386-linux -Dvendorprefix=/usr -Dsiteprefix=/usr -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -march=i386 -mcpu=i686',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/include/gdbm'
    ccversion='', gccversion='3.2 20020822 (Red Hat Linux Rawhide 3.2-5)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.92.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.2.92'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.8.0/i386-linux-thread-multi/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl/5.6.1
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.0
    /usr/lib/perl5/vendor_perl/5.6.1
    /usr/lib/perl5/vendor_perl
    .


Environment for perl v5.8.0:
    HOME=/home/local/swhite
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/home/shared/builds/lib/:/usr/local/games/quake:/usr/local/glide/lib/
    LOGDIR (unset)
    PATH=/usr/java/j2sdk1.4.1/:/usr/java/j2sdk1.4.1/:/usr/java/j2sdk1.4.1/:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/local/scripts:/usr/lib/OpenOffice.org1.0/program/:/usr/local/Acrobat5/bin:/home/local/swhite/bin:/usr/X11R6/bin:/usr/local/scripts:/usr/lib/OpenOffice.org1.0/program/:/usr/local/Acrobat5/bin:/home/local/swhite/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From tassilo.parseval@post.rwth-aachen.de

On Fri\, May 14\, 2004 at 03&#8203;:52&#8203;:55PM \+0000 swhite@&#8203;decisionsoft\.com \(via RT\) wrote&#8203;:

> \# New Ticket Created by  swhite@&#8203;decisionsoft\.com 
> \# Please include the string&#8203;:  \[perl \#29585\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=29585 >
> 
> 
> 
> This is a bug report for perl from swhite@&#8203;decisionsoft\.com\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> The Shell module can't cope with spaces in filenames\.  Take the following code
> 
> use Shell qw\(rm\);
> 
> rm\("my file"\);
> 
> 
> It doesn't work\.  In a fit of cunningness you try&#8203;:
> 
> rm\("'my file'"\);
> 
> But the Shell module nicely escapes the single quotes for you so that doesn't work either\.

That happens because Shell\.pm doesn't escape spaces in command
arguments\, although it should\. Below patch should against blead fix it
\(I suppose there are other characters that need escaping\, too\)\.

Tassilo
\-\- 
$\_=q\#"\,\}\]\)\!JAPH\!qq\(tsuJ\[\{@&#8203;"tnirp\}3\.\.0\}\_$;//&#8203;::niam/s~=\)\]3\[\)\)\_$\-3\(rellac\(=\_$\(\{
pam\{rekcahbus\}\)\(rekcah\{lrePbus\}\)\(lreP\{rehtonabus\}\)\!JAPH\!qq\(rehtona\{tsuJbus\#;
$\_=reverse\,s\+\(?\<=sub\)\.\+q\#q\!'"qq\.\\t$&\."'\!\#\+sexisexiixesixeseg;y~\\n~~dddd;eval

**Comment by p5pRT** [comment]
### From tassilo.parseval@post.rwth-aachen.de

<details><summary>Inline Patch</summary>

```diff
--- lib/Shell.pm~	2004-05-18 06:54:17.000000000 +0200
+++ lib/Shell.pm	2004-05-18 06:57:03.000000000 +0200
@@ -86,7 +86,7 @@ sub AUTOLOAD {
 		    }
 		} else {
 		    for (\@arr) {
-			s/(['\\\\])/\\\\\$1/g;
+			s/(['\\\\ ])/\\\\\$1/g;
 			\$_ = \$_;
 		    }
 		}
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @eserte

Tassilo von Parseval \<tassilo\.parseval@&#8203;post\.rwth\-aachen\.de> wrote&#8203;:
> On Fri\, May 14\, 2004 at 03&#8203;:52&#8203;:55PM \+0000 swhite@&#8203;decisionsoft\.com \(via RT\) wrote&#8203;:
> 
> > \# New Ticket Created by  swhite@&#8203;decisionsoft\.com 
> > \# Please include the string&#8203;:  \[perl \#29585\]
> > \# in the subject line of all future correspondence about this issue\. 
> > \# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=29585 >
> > 
> > 
> > 
> > This is a bug report for perl from swhite@&#8203;decisionsoft\.com\,
> > generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > The Shell module can't cope with spaces in filenames\.  Take the following code
> > 
> > use Shell qw\(rm\);
> > 
> > rm\("my file"\);
> > 
> > 
> > It doesn't work\.  In a fit of cunningness you try&#8203;:
> > 
> > rm\("'my file'"\);
> > 
> > But the Shell module nicely escapes the single quotes for you so that doesn't work either\.
> 
> That happens because Shell\.pm doesn't escape spaces in command
> arguments\, although it should\. Below patch should against blead fix it
> \(I suppose there are other characters that need escaping\, too\)\.
> 
> Tassilo
> \-\- 
> $\_=q\#"\,\}\]\)\!JAPH\!qq\(tsuJ\[\{@&#8203;"tnirp\}3\.\.0\}\_$;//&#8203;::niam/s~=\)\]3\[\)\)\_$\-3\(rellac\(=\_$\(\{
> pam\{rekcahbus\}\)\(rekcah\{lrePbus\}\)\(lreP\{rehtonabus\}\)\!JAPH\!qq\(rehtona\{tsuJbus\#;
> $\_=reverse\,s\+\(?\<=sub\)\.\+q\#q\!'"qq\.\\t$&\."'\!\#\+sexisexiixesixeseg;y~\\n~~dddd;eval
> \-\-\- lib/Shell\.pm~    2004\-05\-18 06&#8203;:54&#8203;:17\.000000000 \+0200
> \+\+\+ lib/Shell\.pm    2004\-05\-18 06&#8203;:57&#8203;:03\.000000000 \+0200
> @&#8203;@&#8203; \-86\,7 \+86\,7 @&#8203;@&#8203; sub AUTOLOAD \{
>              \}
>          \} else \{
>              for \(\\@&#8203;arr\) \{
> \-            s/\(\['\\\\\\\\\]\)/\\\\\\\\\\$1/g;
> \+            s/\(\['\\\\\\\\ \]\)/\\\\\\\\\\$1/g;
>              \\$\_ = \\$\_;
>              \}
>          \}

Shouldn't Shell\.pm rather use the list form of system\(\) resp\. the
safe pipe form open SUBPROC\, "\-|" or exec @&#8203;cmd ?

Regards\,
&nbsp;    Slaven

\-\- 
&nbsp;     \_\_o        Slaven Rezic
&nbsp;   \_\`\\\<\,\_        slaven \<at> rezic \<dot> de
\_\_\(\_\)/ \(\_\)\_\_\_\_
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From tassilo.parseval@post.rwth-aachen.de

On Tue\, May 18\, 2004 at 08&#8203;:22&#8203;:05AM \+0000 slaven@&#8203;rezic\.de wrote&#8203;:

> Tassilo von Parseval \<tassilo\.parseval@&#8203;post\.rwth\-aachen\.de> wrote&#8203;:
> > On Fri\, May 14\, 2004 at 03&#8203;:52&#8203;:55PM \+0000 swhite@&#8203;decisionsoft\.com \(via RT\) wrote&#8203;:

> > > The Shell module can't cope with spaces in filenames\.  Take the following code
> > > 
> > > use Shell qw\(rm\);
> > > 
> > > rm\("my file"\);
> > > 
> > > 
> > > It doesn't work\.  In a fit of cunningness you try&#8203;:
> > > 
> > > rm\("'my file'"\);
> > > 
> > > But the Shell module nicely escapes the single quotes for you so that doesn't work either\.
> > 
> > That happens because Shell\.pm doesn't escape spaces in command
> > arguments\, although it should\. Below patch should against blead fix it
> > \(I suppose there are other characters that need escaping\, too\)\.

> > \-\-\- lib/Shell\.pm~    2004\-05\-18 06&#8203;:54&#8203;:17\.000000000 \+0200
> > \+\+\+ lib/Shell\.pm    2004\-05\-18 06&#8203;:57&#8203;:03\.000000000 \+0200
> > @&#8203;@&#8203; \-86\,7 \+86\,7 @&#8203;@&#8203; sub AUTOLOAD \{
> >              \}
> >          \} else \{
> >              for \(\\@&#8203;arr\) \{
> > \-            s/\(\['\\\\\\\\\]\)/\\\\\\\\\\$1/g;
> > \+            s/\(\['\\\\\\\\ \]\)/\\\\\\\\\\$1/g;
> >              \\$\_ = \\$\_;
> >              \}
> >          \}
> 
> Shouldn't Shell\.pm rather use the list form of system\(\) resp\. the
> safe pipe form open SUBPROC\, "\-|" or exec @&#8203;cmd ?

Better not\. If the list\-form was used\, you could no longer write

&nbsp;    rm\("\*"\);

and expect globbing to happen\.

Tassilo
\-\- 
$\_=q\#"\,\}\]\)\!JAPH\!qq\(tsuJ\[\{@&#8203;"tnirp\}3\.\.0\}\_$;//&#8203;::niam/s~=\)\]3\[\)\)\_$\-3\(rellac\(=\_$\(\{
pam\{rekcahbus\}\)\(rekcah\{lrePbus\}\)\(lreP\{rehtonabus\}\)\!JAPH\!qq\(rehtona\{tsuJbus\#;
$\_=reverse\,s\+\(?\<=sub\)\.\+q\#q\!'"qq\.\\t$&\."'\!\#\+sexisexiixesixeseg;y~\\n~~dddd;eval

**Comment by p5pRT** [comment]
### From @eserte

Tassilo von Parseval \<tassilo\.parseval@&#8203;post\.rwth\-aachen\.de> wrote&#8203;:
> On Tue\, May 18\, 2004 at 08&#8203;:22&#8203;:05AM \+0000 slaven@&#8203;rezic\.de wrote&#8203;:
> 
> > Tassilo von Parseval \<tassilo\.parseval@&#8203;post\.rwth\-aachen\.de> wrote&#8203;:
> > > On Fri\, May 14\, 2004 at 03&#8203;:52&#8203;:55PM \+0000 swhite@&#8203;decisionsoft\.com \(via RT\) wrote&#8203;:
> 
> > > > The Shell module can't cope with spaces in filenames\.  Take the following code
> > > > 
> > > > use Shell qw\(rm\);
> > > > 
> > > > rm\("my file"\);
> > > > 
> > > > 
> > > > It doesn't work\.  In a fit of cunningness you try&#8203;:
> > > > 
> > > > rm\("'my file'"\);
> > > > 
> > > > But the Shell module nicely escapes the single quotes for you so that doesn't work either\.
> > > 
> > > That happens because Shell\.pm doesn't escape spaces in command
> > > arguments\, although it should\. Below patch should against blead fix it
> > > \(I suppose there are other characters that need escaping\, too\)\.
> 
> > > \-\-\- lib/Shell\.pm~    2004\-05\-18 06&#8203;:54&#8203;:17\.000000000 \+0200
> > > \+\+\+ lib/Shell\.pm    2004\-05\-18 06&#8203;:57&#8203;:03\.000000000 \+0200
> > > @&#8203;@&#8203; \-86\,7 \+86\,7 @&#8203;@&#8203; sub AUTOLOAD \{
> > >              \}
> > >          \} else \{
> > >              for \(\\@&#8203;arr\) \{
> > > \-            s/\(\['\\\\\\\\\]\)/\\\\\\\\\\$1/g;
> > > \+            s/\(\['\\\\\\\\ \]\)/\\\\\\\\\\$1/g;
> > >              \\$\_ = \\$\_;
> > >              \}
> > >          \}
> > 
> > Shouldn't Shell\.pm rather use the list form of system\(\) resp\. the
> > safe pipe form open SUBPROC\, "\-|" or exec @&#8203;cmd ?
> 
> Better not\. If the list\-form was used\, you could no longer write
> 
>     rm\("\*"\);
> 
> and expect globbing to happen\.
> 

Then one could argue that rm\("my file"\) is legitimate\, too\, to delete
the files "my" and "file"\.

Regards\,
&nbsp;    Slaven

\-\- 
&nbsp;     \_\_o        Slaven Rezic
&nbsp;   \_\`\\\<\,\_        slaven \<at> rezic \<dot> de
\_\_\(\_\)/ \(\_\)\_\_\_\_
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From wolfgang.laun@alcatel.at

On Tue\, 18 May 2004 07&#8203;:01&#8203;:40 \+0200 Tassilo von Parseval wrote&#8203;:
> On Fri\, May 14\, 2004 at 03&#8203;:52&#8203;:55PM \+0000 swhite@&#8203;decisionsoft\.com \(via RT\) wrote&#8203;:
> > The Shell module can't cope with spaces in filenames\.  Take the following code
> >
> > use Shell qw\(rm\);
> > 
> > rm\("my file"\);
> > 
> > It doesn't work\.  In a fit of cunningness you try&#8203;:
> > 
> > rm\("'my file'"\);
> 
> That happens because Shell\.pm doesn't escape spaces in command
> arguments\, although it should\. Below patch should against blead fix it
> \(I suppose there are other characters that need escaping\, too\)\.
> 

Well\, \\t is an obvious candidate\. But\.\.\.

Quoting ' ' might break current programs\, since e\.g\.
&nbsp;  ls\( "\-l $files"\)
won't work any more\. \(Although it's nowhere in the pod that this
should work\.\)

Before any patching is being done\, here is a quote from
20000236\.008\, aka \[perl \#2725\]&#8203;:

" Consider the example&#8203;: $foo = echo\("howdy"\, "\<funny>"\, "world"\);
" If arguments are not quoted any more\, this results in an error\,
" sh&#8203;: funny&#8203;: No such file or directory
" With quoting\, cat\("\</etc/passwd"\) doesn't work\.

Both examples are still in the Shell\.pm pod\.

" The gist of all of this is\, that the user must be aware of the shell
" seeing the arguments\, and use quoting accordingly\.
"
" WHAT IS THE INTENDED BEHAVIOUR OF THIS MODULE?

At some time selective quoting was added for ' and \\\. Not for "\, so
the original example could have been written as 
&nbsp;    rm\( '"my file"' \);
and the example from the pod as&#8203;:
&nbsp;    echo\("howdy"\, '"\<funny>"'\, "world"\);  \# returns howdy \<funny> world
Again\, nothing is documented\.

Sorry\, I don't think that any change to the code would improve things\.
As for the pod \- that's quite another matter\.

Regards
\-Wolfgang

**Comment by p5pRT** [comment]
### From @demerphq


> Better not\. If the list\-form was used\, you could no longer write
> 
>     rm\("\*"\);
> 
> and expect globbing to happen\.

Which you can't portably do anyway right?

Yves

**Comment by p5pRT** [comment]
### From @eserte

Tassilo von Parseval \<tassilo\.parseval@&#8203;post\.rwth\-aachen\.de> wrote&#8203;:
> On Tue\, May 18\, 2004 at 08&#8203;:22&#8203;:05AM \+0000 slaven@&#8203;rezic\.de wrote&#8203;:
> 
> > Tassilo von Parseval \<tassilo\.parseval@&#8203;post\.rwth\-aachen\.de> wrote&#8203;:
> > > On Fri\, May 14\, 2004 at 03&#8203;:52&#8203;:55PM \+0000 swhite@&#8203;decisionsoft\.com \(via RT\) wrote&#8203;:
> 
> > > > The Shell module can't cope with spaces in filenames\.  Take the following code
> > > > 
> > > > use Shell qw\(rm\);
> > > > 
> > > > rm\("my file"\);
> > > > 
> > > > 
> > > > It doesn't work\.  In a fit of cunningness you try&#8203;:
> > > > 
> > > > rm\("'my file'"\);
> > > > 
> > > > But the Shell module nicely escapes the single quotes for you so that doesn't work either\.
> > > 
> > > That happens because Shell\.pm doesn't escape spaces in command
> > > arguments\, although it should\. Below patch should against blead fix it
> > > \(I suppose there are other characters that need escaping\, too\)\.
> 
> > > \-\-\- lib/Shell\.pm~    2004\-05\-18 06&#8203;:54&#8203;:17\.000000000 \+0200
> > > \+\+\+ lib/Shell\.pm    2004\-05\-18 06&#8203;:57&#8203;:03\.000000000 \+0200
> > > @&#8203;@&#8203; \-86\,7 \+86\,7 @&#8203;@&#8203; sub AUTOLOAD \{
> > >              \}
> > >          \} else \{
> > >              for \(\\@&#8203;arr\) \{
> > > \-            s/\(\['\\\\\\\\\]\)/\\\\\\\\\\$1/g;
> > > \+            s/\(\['\\\\\\\\ \]\)/\\\\\\\\\\$1/g;
> > >              \\$\_ = \\$\_;
> > >              \}
> > >          \}
> > 
> > Shouldn't Shell\.pm rather use the list form of system\(\) resp\. the
> > safe pipe form open SUBPROC\, "\-|" or exec @&#8203;cmd ?
> 
> Better not\. If the list\-form was used\, you could no longer write
> 
>     rm\("\*"\);
> 
> and expect globbing to happen\.
> 

Another thought&#8203;: do it as in system\(\) and exec\(\) and use list form
only if a list was supplied\, otherwise the string form\. There's just
one problem&#8203;: if only one argument was supplied it is not possible
to tell whether to use the list or string form \(this is also a problem
with system\(\) and exec\(\) if there's only one parameter which contains
shell meta characters e\.g\. a program path with a space\)\. To
workaround this\, one could use

&nbsp;    rm\(\["my file"\]\)

or

&nbsp;    rm\("my file"\, undef\)

instead \(I prefer the first\)\.

Regards\,
&nbsp;    Slaven

\-\- 
&nbsp;     \_\_o        Slaven Rezic
&nbsp;   \_\`\\\<\,\_        slaven \<at> rezic \<dot> de
\_\_\(\_\)/ \(\_\)\_\_\_\_
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From tassilo.parseval@post.rwth-aachen.de

On Tue\, May 18\, 2004 at 08&#8203;:54&#8203;:20AM \+0000 slaven@&#8203;rezic\.de wrote&#8203;:

> Tassilo von Parseval \<tassilo\.parseval@&#8203;post\.rwth\-aachen\.de> wrote&#8203;:
> > On Tue\, May 18\, 2004 at 08&#8203;:22&#8203;:05AM \+0000 slaven@&#8203;rezic\.de wrote&#8203;:

> > > Shouldn't Shell\.pm rather use the list form of system\(\) resp\. the
> > > safe pipe form open SUBPROC\, "\-|" or exec @&#8203;cmd ?
> > 
> > Better not\. If the list\-form was used\, you could no longer write
> > 
> >     rm\("\*"\);
> > 
> > and expect globbing to happen\.
> > 
> 
> Then one could argue that rm\("my file"\) is legitimate\, too\, to delete
> the files "my" and "file"\.

Hmmh\, right\, this is tricky\. The current behaviour \(that is\, globbing
happens\, for instance\) suggests\, that Shell\.pm simply passes all
arguments unaltered to the shell and should not use the list\-form for
that reason\. If I were to write&#8203;:

&nbsp;    rm\( q\{my\\ file\} \);

then it should expand to something like 

&nbsp;    bash \-c 'rm my\\ file'

Right now however\, it is very inconsistent because Shell\.pm escapes
certain characters \('\\' is one of them\) and leaves others \(e\.g\. '\*'\)
untouched so my above attempt to delete "my file" fails with

&nbsp;    rm&#8203;: cannot remove \`my\\\\'&#8203;: No such file or directory
&nbsp;    rm&#8203;: cannot remove \`file'&#8203;: No such file or directory

So I think the right thing to do would be to have it use list\-form if a
list was provided\, otherwise invoke the shell to allow meta\-characters\.
That way I could write both

&nbsp;    rm\( glob "\*" \);

or \(safer because glob\(\) might only return one value\)

&nbsp;    rm\( \[ glob "\*" \] \);

and 

&nbsp;    rm\("\*"\);

Unfortunately\, this will prevent anyone from writing

&nbsp;    cp\("\*"\, "/tmp"\);

In short&#8203;: There are many possible ways of fixing Shell\.pm\. But each of
them leaves quite a lot to be desired\.

I am inclined to think that Shell\.pm was a bad idea and that it
shouldn't have made it into the core at all\.

Tassilo
\-\- 
$\_=q\#"\,\}\]\)\!JAPH\!qq\(tsuJ\[\{@&#8203;"tnirp\}3\.\.0\}\_$;//&#8203;::niam/s~=\)\]3\[\)\)\_$\-3\(rellac\(=\_$\(\{
pam\{rekcahbus\}\)\(rekcah\{lrePbus\}\)\(lreP\{rehtonabus\}\)\!JAPH\!qq\(rehtona\{tsuJbus\#;
$\_=reverse\,s\+\(?\<=sub\)\.\+q\#q\!'"qq\.\\t$&\."'\!\#\+sexisexiixesixeseg;y~\\n~~dddd;eval

**Comment by p5pRT** [comment]
### From @rgs

Tassilo von Parseval wrote&#8203;:
> 
> I am inclined to think that Shell\.pm was a bad idea and that it
> shouldn't have made it into the core at all\.

I'm inclined to think that Shell\.pm was mostly a funny toy and that it
shouldn't be used seriously\.

**Comment by p5pRT** [comment]
### From swhite@decisionsoft.com

\-\-\-\- Original Message \-\-\-\-
> From Rafael Garcia\-Suarez via RT \<perlbug\-followup@&#8203;perl\.org>
> Date&#8203;: Wednesday\, 19 May 2004\, 08&#8203;:25
>
> I'm inclined to think that Shell\.pm was mostly a funny toy and that it
> shouldn't be used seriously\.

Yes\, this doesn't help when you come to fix bugs in code that uses it
though\.  As pointed out previously the POD doesn't discourage it's use
and it is ambigious in explaining how to use it WRT to quoting\.

Given it's supposed to emulate shell scripting the option of quoting
nothing seems most consistant\, leave that to the user\, but this isn't
backwards compatible given the current implementation which does quote
some characters\.  If people want a function that deals with quoting
everything they're going about things the wrong way and should be using
the array form of 'system'\.

One co\-worker suggested adding an 'escapechars' option to the module\, so 
you can import it with something like 
'use Shell qw/&#8203;:escapechars=>"" rm/;

Whether this is worth the effort I'm not sure\.  Maybe the POD should
just state at the top that Shell\.pm is a cool autoloader toy/demo and
real code should use system or other alternative\.

\-\- 
Stephen White                               \+44\-1865\-203192
DecisionSoft Limited                        http&#8203;://www\.decisionsoft\.com
XML Development and Services

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
